### PR TITLE
Complex-valued variables

### PR DESCRIPTION
**Pull request by projekter**
This is in conjuction with https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/pull/213 and provides the DynamicPolynomials implementation of the complex-valued variable functionality.

### CODE REVIEW

**Comment by projekter** [comment]
Issue with sometimes working implementation for incomplete substitution (I just quote the code comment)
> We may assign a complex or real variable to its value (ordinary substitution).
> We may also assign a complex value to its conjugate, or just the real or imaginary parts
> Any combination of z, conj(z), real(z), imag(z), imag(conj(z)) can occur in either the polynomial or the substitution,
> and we must handle all of them correctly.
> Note: This may or may not work... Issues can arise if the substitutions contain the real and imaginary (or only one of
> those) of a variable separately whenever vals is not of the correct type:
> - Unless subs() originally had a polynomial-valued rhs, vals will be scalars, monomials, or terms. So when we try to
>   assign a polynomial to its value (which is necessary, as the one-step substitution of only the real or only the
>   imaginary part is incomplete), this is an impossible conversion.
> - The coefficients in vals might not be complex-valued; but to assign only a part of the variable, we necessarily need to
>   introduce an explicit imaginary coefficient to the value.
> Currently, we don't do anything to catch these errors.

Design question: With the new implementation, the "dynamic" part becomes slightly more "typed", as the variable and monomial ordering are now also part of the type. Still, the complex kind is a field instead of a type parameter. While I definitely would keep this to distinguish between the variable, its conjugate, its real and imaginary part (else all containers holding such variables would need to be abstract containers, making this very inefficient), it may be debatable whether the issue of "is this a complex type at all" should be encoded as a type parameter. Pro: Some replacements may be simpler to decide based on static dispatch. Con: Duplication of information; and mixing real and complex variables again has to fall back to a generic implementation.

File: src/mono.jl [hunk 1]
```diff
@@ -100,3 +100,10 @@ MP.monomial(m::Monomial) = m
 #    end
 #    i > length(m1.z)
 #end
+
+# for efficiency reasons
+function Base.conj(x::Monomial{true})
```
Comment (blegat) [inline src/mono.jl:105]: `{true}` is still the old syntax

**Comment by blegat** [comment]
Yes, I would definitely not incorporate whether it's a real, imaginary part etc... as type as it would disallow mixing different ones in a polynomial etc...
About whether there should be a type for saying whether it was created with `@polyvar` and `@polycvar`, that would disallow mixing the two types of variables in the same polynomial for instance so that wouldn't be ideal right ?

**Comment by projekter** [comment]
This would depend on the actual implementation in `Monomial`. Combining both would still work if the `vars` vectors still only contained elements of type `Variable{V,M}`, not specializing the concrete complex type `C`. But then, this would be horribly inefficient as Julia would have to generate type checks and dynamic dispatches on every operation. So I agree this would be very "not ideal".

File: src/var.jl [hunk 2]
```diff
@@ -114,19 +143,37 @@ function instantiate(::Type{NonCommutative{O}}) where {O}
     return NonCommutative(instantiate(O))
 end
 
+@enum ComplexKind cpNone cpFull cpConj cpReal cpImag
```
Comment (blegat) [inline src/var.jl:146]: Could you add a docstring here ?

Comment (projekter) [reply src/var.jl:163]: Done - should this then be exported?

Comment (blegat) [reply src/var.jl:163]: Yes, I like to follow https://jump.dev/JuMP.jl/stable/developers/style/#Exported-and-non-exported-names

Comment (blegat) [reply src/var.jl:163]: Can we rename the constants with capital letters ? What about : `REAL`, `COMPLEX`, `CONJ`, `REAL_PART`, `IMAG_PART` ? `CONJ` and `IMAG` could also be expanded but this is consistent with `Base.conj` and `Base.imag`

Comment (blegat) [reply src/var.jl:163]: Yes, it's ok to not export then.

> Documenter doesn't like non-exported docstrings.

What do you mean ? DynamicPolynomials doesn't have a documentation anyway.

> With the prefix (which unfortunately isn't so widespread for enums outside of Pascal-like languages) at least the danger of collisions is reduced.

I'm not a big fan of prefix like `cp` which reduces readability. I prefer non-exported long names like `DynamicPolynomials.REAL` than exported short names with a prefix to try to avoid collision. The way we deal with collision in the JuMP style guide is that we advise to use `import` but we don't have to follow this guide strictly so let's not export and have short, simple names without cryptic prefix

Comment (projekter) [reply src/var.jl]: Ok, renamed, but kept unexported.
> What do you mean ? 

If I remember correctly, Documenter warns (fails if not excepted) if you have a method with a docstring that is not exported. Perhaps it also was a method with a docstring that is not included in the documentation, I don't remember. Which kind of makes sense: A docstring is for user documentation, if you want developer documentation, use a comment (maybe).

Well, prefixes are good if there's no way to do scoping in the language. With the modularization in Julia, this is already better as explicit package prefixes are now required. Still, a clean naming of enums IMHO means that the name of the enum value makes clear where it belongs to. This is possible is the language offers/requires you to write `Enumtype.enumvalue`, but since we all like expressive names, `Enumtype` might not be very short, so this unnecessarily makes both writing and reading harder. Short expressive prefixes are a very good compromise that also help for code completion without the full might of prefixing by the enum name.

**Comment by blegat** [comment]
@projekter I forgot about this PR, we should merge it

**Comment by projekter** [comment]
@blegat Here's a reminder

File: src/DynamicPolynomials.jl [hunk 3]
```diff
@@ -72,11 +72,11 @@ function MP.similar_variable(
...
-function MP.similar_variable(
-    ::Union{PolyType{V,M},Type{<:PolyType{V,M}}},
-    s::Symbol,
-) where {V,M}
-    return Variable(string(s), V, M)
+function MP.similar_variable(p::PolyType{V,M}, s::Symbol) where {V,M}
+    return Variable(string(s), V, M, isreal(p) ? REAL : COMPLEX)
+end
+function MP.similar_variable(::Type{<:PolyType{V,M}}, s::Symbol) where {V,M}
+    return Variable(string(s), V, M, REAL) # we cannot infer this from the type,
```
Comment (blegat) [inline src/DynamicPolynomials.jl:79]:
```suggestion
    return Variable(string(s), V, M, REAL) # we cannot infer this from the type
```

File: src/subs.jl [hunk 4]
```diff
@@ -9,23 +9,73 @@ end
...
+                    else
+                        vals[j] = imag(value)
+                    end
+                elseif s.first.kind == REAL_PART
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the real part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == COMPLEX
+                        vals[j] = value + im * imag(vals[j])
```
Comment (blegat) [inline src/subs.jl:55]: What if `vals[j]` is not assigned yet ? Maybe we can just error in these cases, we can always revisit later

File: src/subs.jl [hunk 5]
```diff
@@ -9,23 +9,73 @@ end
...
+    # Any combination of z, conj(z), real(z), imag(z), imag(conj(z)) can occur in either the polynomial or the substitution,
+    # and we must handle all of them correctly.
+    # Note: This may or may not work... Issues can arise if the substitutions contain the real and imaginary (or only one of
+    # those) of a variable separately whenever vals is not of the correct type:
+    # - Unless subs() originally had a polynomial-valued rhs, vals will be scalars, monomials, or terms. So when we try to
+    #   assign a polynomial to its value (which is necessary, as the one-step substitution of only the real or only the
+    #   imaginary part is incomplete), this is an impossible conversion.
+    # - The coefficients in vals might not be complex-valued; but to assign only a part of the variable, we necessarily need to
+    #   introduce an explicit imaginary coefficient to the value.
+    # Currently, we don't do anything to catch these errors.
```
Comment (blegat) [inline src/subs.jl:26]: I'd rather have a solid implementation that only handle some cases ans error in other cases than an implementation that does not error but that may have incorrect behavior.

Comment (projekter) [reply src/subs.jl:26]: Ok, now I modified it to a different implementation. Basically, real and imaginary parts, when occurring in the substitution rules, are seen as assignments to real values, so they won't act on the complex variables at all, if those occur. You could then say that this (call it lexicographic substitution) is inconsistent with the complex variable in the rule still acting on the real and imaginary parts, and I agree. However, where to draw the line? It is then also inconsistent to apply it to the conjugate, but this is (I think) a very reasonable choice.
I guess that a typical situation will probably never mix real/imaginary part variables with full complex variables, so I don't expect whatever behavior is defined here to actually be used somewhere.

File: src/subs.jl (hunk 4 above)
Comment (projekter) [reply src/subs.jl:55]: Now only full substitution is done, so the problem doesn't occur any more.

File: src/subs.jl [hunk 6]
```diff
@@ -9,23 +9,63 @@ end
...
+                else
+                    @assert(s.first.kind == IMAG_PART)
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the imaginary part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == IMAG_PART
+                        vals[j] = value
+                    end
+                    # don't do a partial substitution
```
Comment (blegat) [inline src/subs.jl:65]: I think we should error here. As you said in your comment:

> I guess that a typical situation will probably never mix real/imaginary part variables with full complex variables, so I don't expect whatever behavior is defined here to actually be used somewhere.

I agree, for this reason, an error is fine because when this is used, this will usually be a mistake in which case an error is helpful. On the other hand, if we silently ignore here, then it will not be helpful debugging these mistakes.

Comment (projekter) [reply src/subs.jl]: Ah, understood. I thought that the previous change is well-defined behavior, but of course it is not documented how substition should work in these cases in the MultivariatePolynomials interface. Now it is an error.

File: src/subs.jl [hunk 7]
```diff
@@ -9,23 +9,69 @@ end
...
+                    @assert(s.first.kind == IMAG_PART)
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the imaginary part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == IMAG_PART
+                        vals[j] = value
+                    elseif vars[j].kind != REAL_PART
+                        error(
+                            "Found complex variable with substitution of imaginary part - not implemented",
```
Comment (blegat) [inline src/subs.jl:69]: Nice, could you add some tests for these edge cases ?

Comment (projekter) [reply src/subs.jl:69]: I added a few testcases. Note that I did not use the substitution syntax without explicit variable specification. This is because `variables(p)` would list all variables, complex, conjugates, real, and imaginary parts separately. So you'd have to specify values for all of them (unnecessarily - and this would even be more restrictive as then: `z + real(z)`, called with arguments `(1+2im, 1)` would trigger the new error, because `z` occurs but a value is assigned to `real(z)`). Either the whole variable assignment would need to be redesigned or we just say that for complex variables, you should always explicitly specify a substitution rule.
(But even then, substitution rules could be redundant; `[z, real(z)] => [1+2im, 1]` is also a valid rule. Should we check whether the rules are self-consistent? Given that variables occuring twice are also not checked, I think the current lax behavior is fine.

Comment (blegat) [reply src/subs.jl:69]: Indeed, that's an issue. We should only allow `COMPLEX_PART` to used in substitutions then. In this PR, I think we should target the simplest implementation that's bullet proof. We can try to make something more complicated but I'd rather do it in a separate PR if there is a use case for it. So I'd just restrict `z` to be a `COMPLEX_PART` whenever `z => ...` is used

Comment (projekter) [reply src/subs.jl]: This simplifies the substitution process significantly. I don't know whether I like it, but there's some sense to it.

Comment (blegat) [reply src/subs.jl]: We can revisit in a follow up PR without being breaking. We'll be able to recover your implementation easily from this branch.
